services/director-v2/src/simcore_service_director_v2/modules/db/repositories/clusters.py

```diff
@@ -71,11 +90,201 @@ async def _clusters_from_cluster_ids(
             if not clusters_list:
                 raise ClusterNotFoundError(cluster_id=cluster_id)
-            logger.debug("found cluster in DB: %s", f"{clusters_list[0]=}")
-            return clusters_list[0]
+            the_cluster = clusters_list[0]
+
+            access_rights = await _compute_user_access_rights(
+                conn, user_id, the_cluster
+            )
+            logger.debug(
+                "found cluster in DB: %s, with computed %s",
+                f"{the_cluster=}",
+                f"{access_rights=}",
+            )
+        if not access_rights.read:
+            raise ClusterAccessForbiddenError(cluster_id=cluster_id)
+
+        return the_cluster
+
+    async def update_cluster(  # pylint: disable=too-many-branches
+        self, user_id: UserID, cluster_id: ClusterID, updated_cluster: ClusterPatch
+    ) -> Cluster:
+        async with self.db_engine.acquire() as conn:
+            clusters_list = await _clusters_from_cluster_ids(conn, {cluster_id})
+            if not clusters_list:
+                raise ClusterNotFoundError(cluster_id=cluster_id)
+            the_cluster = clusters_list[0]
+
+            this_user_access_rights = await _compute_user_access_rights(
+                conn, user_id, the_cluster
+            )
+            logger.debug(
+                "found cluster in DB: %s, with computed %s",
+                f"{the_cluster=}",
+                f"{this_user_access_rights=}",
+            )
+
+            if not this_user_access_rights.write:
+                raise ClusterAccessForbiddenError(cluster_id=cluster_id)
+
+            if updated_cluster.owner and updated_cluster.owner != the_cluster.owner:
+                # if the user wants to change the owner, we need more rights here
+                if this_user_access_rights != CLUSTER_ADMIN_RIGHTS:
+                    raise ClusterAccessForbiddenError(cluster_id=cluster_id)
+
+                # ensure the new owner has admin rights, too
+                if not updated_cluster.access_rights:
+                    updated_cluster.access_rights = {
+                        updated_cluster.owner: CLUSTER_ADMIN_RIGHTS
+                    }
+                else:
+                    updated_cluster.access_rights[
+                        updated_cluster.owner
+                    ] = CLUSTER_ADMIN_RIGHTS
+
+            # resolve access rights changes
+            resolved_access_rights = the_cluster.access_rights
+            if updated_cluster.access_rights:
+                # if the user is a manager he/she may ONLY add/remove users
+                if this_user_access_rights == CLUSTER_MANAGER_RIGHTS:
+                    for grp, rights in updated_cluster.access_rights.items():
+                        if grp in [the_cluster.owner] or rights not in [
+                            CLUSTER_USER_RIGHTS,
+                            CLUSTER_NO_RIGHTS,
+                        ]:
```

Comment:
Maybe have it like this? Especially the first check should be way better.
```suggestion
                        if grp == the_cluster.owner or rights not in {
                            CLUSTER_USER_RIGHTS,
                            CLUSTER_NO_RIGHTS,
                        }:
```